File: android/app/src/main/java/com/foreverjukebox/app/engine/Graph.kt
package com.foreverjukebox.app.engine

private const val TIMBRE_WEIGHT = 1.0
private const val PITCH_WEIGHT = 10.0
private const val LOUD_START_WEIGHT = 1.0
private const val LOUD_MAX_WEIGHT = 1.0
private const val DURATION_WEIGHT = 100.0
private const val CONFIDENCE_WEIGHT = 1.0
private const val MAX_DISTANCE = 100.0
private const val FULL_MATCH_DISTANCE = 0.0
private const val TARGET_BRANCH_DIVISOR = 6
private const val THRESHOLD_START = 10
private const val THRESHOLD_STEP = 5
private const val LONGEST_BACKWARD_THRESHOLD = 50
private const val ADD_LAST_EDGE_HIGH_THRESHOLD = 65
private const val ADD_LAST_EDGE_LOW_THRESHOLD = 55
private const val REACH_THRESHOLD = 50

private fun euclideanDistance(v1: List<Double>, v2: List<Double>): Double {
    var sum = 0.0
    for (i in v1.indices) {
        val delta = v2[i] - v1[i]
        sum += delta * delta
    }
    return kotlin.math.sqrt(sum)
}

private fun segmentDistance(seg1: Segment, seg2: Segment): Double {
    val timbre = euclideanDistance(seg1.timbre, seg2.timbre)
    val pitch = euclideanDistance(seg1.pitches, seg2.pitches)
    val loudStart = kotlin.math.abs(seg1.loudnessStart - seg2.loudnessStart)
    val loudMax = kotlin.math.abs(seg1.loudnessMax - seg2.loudnessMax)
    val duration = kotlin.math.abs(seg1.duration - seg2.duration)
    val confidence = kotlin.math.abs(seg1.confidence - seg2.confidence)
    return timbre * TIMBRE_WEIGHT +
        pitch * PITCH_WEIGHT +
        loudStart * LOUD_START_WEIGHT +
        loudMax * LOUD_MAX_WEIGHT +
        duration * DURATION_WEIGHT +
        confidence * CONFIDENCE_WEIGHT
}

private fun calculateNearestNeighborsForQuantum(
    quanta: List<QuantumBase>,
    maxNeighbors: Int,
    maxThreshold: Int,
    q1: QuantumBase,
    allEdges: MutableList<Edge>
) {
    val edges = mutableListOf<Edge>()
    if (q1.overlappingSegments.isEmpty()) {
        q1.allNeighbors = mutableListOf()
        return
    }

    for (i in quanta.indices) {
        if (i == q1.which) continue
        val q2 = quanta[i]
        var sum = 0.0
        for (j in q1.overlappingSegments.indices) {
            val seg1 = q1.overlappingSegments[j]
            val distance = if (j < q2.overlappingSegments.size) {
                val seg2 = q2.overlappingSegments[j]
                if (seg1.which == seg2.which) MAX_DISTANCE else segmentDistance(seg1, seg2)
            } else {
                MAX_DISTANCE
            }
            sum += distance
        }

        val pdistance = if (
            q1.indexInParent != null &&
            q2.indexInParent != null &&
            q1.indexInParent == q2.indexInParent
        ) {
            FULL_MATCH_DISTANCE
        } else {
            MAX_DISTANCE
        }

        val totalDistance = sum / q1.overlappingSegments.size + pdistance
        if (totalDistance < maxThreshold) {
            edges.add(
                Edge(
                    id = -1,
                    src = q1,
                    dest = q2,
                    distance = totalDistance,
                    deleted = false
                )
            )
        }
    }

    edges.sortWith(compareBy({ it.distance }))

    q1.allNeighbors = mutableListOf()
    for (i in 0 until kotlin.math.min(maxNeighbors, edges.size)) {
        val edge = edges[i]
        edge.id = allEdges.size
        allEdges.add(edge)
        q1.allNeighbors.add(edge)
    }
}

private fun precalculateNearestNeighbors(
    quanta: List<QuantumBase>,
    maxNeighbors: Int,
    maxThreshold: Int,
    allEdges: MutableList<Edge>
) {
    if (quanta.isEmpty()) return
    if (quanta[0].allNeighbors.isNotEmpty()) return
    allEdges.clear()
    for (q in quanta) {
        calculateNearestNeighborsForQuantum(quanta, maxNeighbors, maxThreshold, q, allEdges)
    }
}

private fun extractNearestNeighbors(
    q: QuantumBase,
    maxThreshold: Int,
    config: JukeboxConfig
): MutableList<Edge> {
    val neighbors = mutableListOf<Edge>()
    for (neighbor in q.allNeighbors) {
        if (neighbor.deleted) continue
        if (config.justBackwards && neighbor.dest.which > q.which) continue
        if (config.justLongBranches &&
            kotlin.math.abs(neighbor.dest.which - q.which) < config.minLongBranch) continue
        if (neighbor.distance <= maxThreshold) {
            neighbors.add(neighbor)
        }
    }
    return neighbors
}

private fun collectNearestNeighbors(
    quanta: List<QuantumBase>,
    maxThreshold: Int,
    config: JukeboxConfig
): Int {
    var branchingCount = 0
    for (q in quanta) {
        q.neighbors = extractNearestNeighbors(q, maxThreshold, config)
        if (q.neighbors.isNotEmpty()) {
            branchingCount += 1
        }
    }
    return branchingCount
}

private fun longestBackwardBranch(quanta: List<QuantumBase>): Double {
    var longest = 0
    for (i in quanta.indices) {
        val q = quanta[i]
        for (neighbor in q.neighbors) {
            val delta = i - neighbor.dest.which
            if (delta > longest) longest = delta
        }
    }
    return (longest * 100.0) / quanta.size
}

private fun insertBestBackwardBranch(
    quanta: List<QuantumBase>,
    threshold: Int,
    maxThreshold: Int
) {
    val branches = mutableListOf<Triple<Double, QuantumBase, Edge>>()
    for (i in quanta.indices) {
        val q = quanta[i]
        for (neighbor in q.allNeighbors) {
            if (neighbor.deleted) continue
            val delta = i - neighbor.dest.which
            if (delta > 0 && neighbor.distance < maxThreshold) {
                val percent = (delta * 100.0) / quanta.size
                branches.add(Triple(percent, q, neighbor))
            }
        }
    }
    if (branches.isEmpty()) return
    branches.sortBy { it.first }
    val best = branches.last()
    if (best.third.distance > threshold) {
        best.second.neighbors.add(best.third)
    }
}

private fun calculateReachability(quanta: List<QuantumBase>) {
    val maxIter = 1000
    for (q in quanta) {
        q.reach = quanta.size - q.which
    }
    for (iter in 0 until maxIter) {
        var changeCount = 0
        for (qi in quanta.indices) {
            val q = quanta[qi]
            var changed = false
            for (neighbor in q.neighbors) {
                val q2 = neighbor.dest
                val q2Reach = q2.reach ?: continue
                val qReach = q.reach ?: continue
                if (q2Reach > qReach) {
                    q.reach = q2Reach
                    changed = true
                }
            }
            if (qi < quanta.size - 1) {
                val q2 = quanta[qi + 1]
                val q2Reach = q2.reach ?: continue
                val qReach = q.reach ?: continue
                if (q2Reach > qReach) {
                    q.reach = q2Reach
                    changed = true
                }
            }
            if (changed) {
                changeCount += 1
                for (j in 0 until q.which) {
                    val q2 = quanta[j]
                    val q2Reach = q2.reach ?: continue
                    val qReach = q.reach ?: continue
                    if (q2Reach < qReach) {
                        q2.reach = qReach
                    }
                }
            }
        }
        if (changeCount == 0) break
    }
}

private fun maxBackwardEdge(q: QuantumBase): Int {
    var maxBackward = 0
    for (neighbor in q.neighbors) {
        val delta = q.which - neighbor.dest.which
        if (delta > maxBackward) {
            maxBackward = delta
        }
    }
    return maxBackward
}

private fun findBestLastBeat(
    quanta: List<QuantumBase>,
    config: JukeboxConfig
): Pair<Int, Double> {
    var longest = 0
    var longestReach = 0.0
    var bestLongIndex = -1
    var bestLongBack = 0
    var bestLongReach = 0.0
    for (i in quanta.size - 1 downTo 0) {
        val q = quanta[i]
        val distanceToEnd = quanta.size - i
        val reach = q.reach?.let { ((it - distanceToEnd) * 100.0) / quanta.size } ?: 0.0
        if (reach > longestReach && q.neighbors.isNotEmpty()) {
            longestReach = reach
            longest = i
        }
        val maxBackward = maxBackwardEdge(q)
        if (q.neighbors.isNotEmpty() && maxBackward >= config.minLongBranch) {
            if (i > bestLongIndex) {
                bestLongIndex = i
                bestLongBack = maxBackward
                bestLongReach = reach
            } else if (i == bestLongIndex) {
                if (maxBackward > bestLongBack || (maxBackward == bestLongBack && reach > bestLongReach)) {
                    bestLongBack = maxBackward
                    bestLongReach = reach
                }
            }
        }
    }
    return if (bestLongIndex >= 0) {
        bestLongIndex to bestLongReach
    } else {
        longest to longestReach
    }
}

private fun filterOutBadBranches(quanta: List<QuantumBase>, lastIndex: Int) {
    for (i in 0 until lastIndex) {
        val q = quanta[i]
        q.neighbors = q.neighbors.filter { it.dest.which < lastIndex }.toMutableList()
    }
}

private fun hasSequentialBranch(q: QuantumBase, neighbor: Edge, lastBranchPoint: Int): Boolean {
    if (q.which == lastBranchPoint) return false
    val qp = q.prev ?: return false
    val distance = q.which - neighbor.dest.which
    for (prevNeighbor in qp.neighbors) {
        val odistance = qp.which - prevNeighbor.dest.which
        if (distance == odistance) return true
    }
    return false
}

private fun filterOutSequentialBranches(quanta: List<QuantumBase>, lastBranchPoint: Int) {
    for (i in quanta.size - 1 downTo 1) {
        val q = quanta[i]
        q.neighbors = q.neighbors.filter { !hasSequentialBranch(q, it, lastBranchPoint) }
            .toMutableList()
    }
}

fun buildJumpGraph(analysis: TrackAnalysis, config: JukeboxConfig): JukeboxGraphState {
    val quanta = analysis.beats
    val allEdges = mutableListOf<Edge>()
    precalculateNearestNeighbors(quanta, config.maxBranches, config.maxBranchThreshold, allEdges)

    var threshold = config.currentThreshold
    if (threshold == 0) {
        val targetBranchCount = quanta.size / TARGET_BRANCH_DIVISOR
        var t = THRESHOLD_START
        while (t < config.maxBranchThreshold) {
            val count = collectNearestNeighbors(quanta, t, config)
            if (count >= targetBranchCount) {
                threshold = t
                break
            }
            t += THRESHOLD_STEP
        }
    }

    if (threshold == 0) threshold = config.maxBranchThreshold

    collectNearestNeighbors(quanta, threshold, config)

    if (config.addLastEdge) {
        if (longestBackwardBranch(quanta) < LONGEST_BACKWARD_THRESHOLD) {
            insertBestBackwardBranch(quanta, threshold, ADD_LAST_EDGE_HIGH_THRESHOLD)
        } else {
            insertBestBackwardBranch(quanta, threshold, ADD_LAST_EDGE_LOW_THRESHOLD)
        }
    }

    calculateReachability(quanta)
    val (lastBranchPoint, longestReach) = findBestLastBeat(quanta, config)
    filterOutBadBranches(quanta, lastBranchPoint)
    if (config.removeSequentialBranches) {
        filterOutSequentialBranches(quanta, lastBranchPoint)
    }

    return JukeboxGraphState(
        computedThreshold = threshold,
        currentThreshold = threshold,
        lastBranchPoint = lastBranchPoint,
        totalBeats = quanta.size,
        longestReach = longestReach,
        allEdges = allEdges
    )
}

File: web/src/engine/graph.ts
import {
  Edge,
  JukeboxConfig,
  JukeboxGraphState,
  QuantumBase,
  Segment,
  TrackAnalysis,
} from "./types";

const TIMBRE_WEIGHT = 1;
const PITCH_WEIGHT = 10;
const LOUD_START_WEIGHT = 1;
const LOUD_MAX_WEIGHT = 1;
const DURATION_WEIGHT = 100;
const CONFIDENCE_WEIGHT = 1;

function euclideanDistance(v1: number[], v2: number[]): number {
  let sum = 0;
  for (let i = 0; i < v1.length; i += 1) {
    const delta = v2[i] - v1[i];
    sum += delta * delta;
  }
  return Math.sqrt(sum);
}

function getSegmentDistance(seg1: Segment, seg2: Segment): number {
  const timbre = euclideanDistance(seg1.timbre, seg2.timbre);
  const pitch = euclideanDistance(seg1.pitches, seg2.pitches);
  const loudStart = Math.abs(seg1.loudness_start - seg2.loudness_start);
  const loudMax = Math.abs(seg1.loudness_max - seg2.loudness_max);
  const duration = Math.abs(seg1.duration - seg2.duration);
  const confidence = Math.abs(seg1.confidence - seg2.confidence);
  return (
    timbre * TIMBRE_WEIGHT +
    pitch * PITCH_WEIGHT +
    loudStart * LOUD_START_WEIGHT +
    loudMax * LOUD_MAX_WEIGHT +
    duration * DURATION_WEIGHT +
    confidence * CONFIDENCE_WEIGHT
  );
}

function calculateNearestNeighborsForQuantum(
  quanta: QuantumBase[],
  maxNeighbors: number,
  maxThreshold: number,
  q1: QuantumBase,
  allEdges: Edge[],
) {
  const edges: Edge[] = [];
  if (q1.overlappingSegments.length === 0) {
    q1.allNeighbors = [];
    return;
  }

  for (let i = 0; i < quanta.length; i += 1) {
    if (i === q1.which) {
      continue;
    }

    const q2 = quanta[i];
    let sum = 0;
    for (let j = 0; j < q1.overlappingSegments.length; j += 1) {
      const seg1 = q1.overlappingSegments[j];
      let distance = 100;
      if (j < q2.overlappingSegments.length) {
        const seg2 = q2.overlappingSegments[j];
        if (seg1.which === seg2.which) {
          distance = 100;
        } else {
          distance = getSegmentDistance(seg1, seg2);
        }
      }
      sum += distance;
    }

    const pdistance =
      q1.indexInParent !== undefined &&
      q2.indexInParent !== undefined &&
      q1.indexInParent === q2.indexInParent
        ? 0
        : 100;

    const totalDistance = sum / q1.overlappingSegments.length + pdistance;
    if (totalDistance < maxThreshold) {
      edges.push({
        id: -1,
        src: q1,
        dest: q2,
        distance: totalDistance,
        deleted: false,
      });
    }
  }

  edges.sort((a, b) =>
    a.distance > b.distance ? 1 : a.distance < b.distance ? -1 : 0,
  );

  q1.allNeighbors = [];
  for (let i = 0; i < maxNeighbors && i < edges.length; i += 1) {
    const edge = edges[i];
    edge.id = allEdges.length;
    allEdges.push(edge);
    q1.allNeighbors.push(edge);
  }
}

function precalculateNearestNeighbors(
  quanta: QuantumBase[],
  maxNeighbors: number,
  maxThreshold: number,
  allEdges: Edge[],
) {
  if (quanta.length === 0) {
    return;
  }
  if (quanta[0].allNeighbors.length > 0) {
    return;
  }
  allEdges.length = 0;
  for (const q of quanta) {
    calculateNearestNeighborsForQuantum(
      quanta,
      maxNeighbors,
      maxThreshold,
      q,
      allEdges,
    );
  }
}

function extractNearestNeighbors(
  q: QuantumBase,
  maxThreshold: number,
  config: JukeboxConfig,
): Edge[] {
  const neighbors: Edge[] = [];
  for (const neighbor of q.allNeighbors) {
    if (neighbor.deleted) {
      continue;
    }
    if (config.justBackwards && neighbor.dest.which > q.which) {
      continue;
    }
    if (
      config.justLongBranches &&
      Math.abs(neighbor.dest.which - q.which) < config.minLongBranch
    ) {
      continue;
    }
    if (neighbor.distance <= maxThreshold) {
      neighbors.push(neighbor);
    }
  }
  return neighbors;
}

function collectNearestNeighbors(
  quanta: QuantumBase[],
  maxThreshold: number,
  config: JukeboxConfig,
): number {
  let branchingCount = 0;
  for (const q of quanta) {
    q.neighbors = extractNearestNeighbors(q, maxThreshold, config);
    if (q.neighbors.length > 0) {
      branchingCount += 1;
    }
  }
  return branchingCount;
}

function longestBackwardBranch(quanta: QuantumBase[]): number {
  let longest = 0;
  for (let i = 0; i < quanta.length; i += 1) {
    const q = quanta[i];
    for (const neighbor of q.neighbors) {
      const delta = i - neighbor.dest.which;
      if (delta > longest) {
        longest = delta;
      }
    }
  }
  return (longest * 100) / quanta.length;
}

function insertBestBackwardBranch(
  quanta: QuantumBase[],
  threshold: number,
  maxThreshold: number,
) {
  const branches: Array<[number, QuantumBase, Edge]> = [];
  for (let i = 0; i < quanta.length; i += 1) {
    const q = quanta[i];
    for (const neighbor of q.allNeighbors) {
      if (neighbor.deleted) {
        continue;
      }
      const delta = i - neighbor.dest.which;
      if (delta > 0 && neighbor.distance < maxThreshold) {
        const percent = (delta * 100) / quanta.length;
        branches.push([percent, q, neighbor]);
      }
    }
  }
  if (branches.length === 0) {
    return;
  }
  branches.sort((a, b) => a[0] - b[0]);
  branches.reverse();
  const [_, bestQ, bestNeighbor] = branches[0];
  if (bestNeighbor.distance > threshold) {
    bestQ.neighbors.push(bestNeighbor);
  }
}

function calculateReachability(quanta: QuantumBase[]) {
  const maxIter = 1000;
  for (const q of quanta) {
    q.reach = quanta.length - q.which;
  }
  // Propagate the furthest reachable beat through backward links and neighbors.
  for (let iter = 0; iter < maxIter; iter += 1) {
    let changeCount = 0;
    for (let qi = 0; qi < quanta.length; qi += 1) {
      const q = quanta[qi];
      let changed = false;
      for (const neighbor of q.neighbors) {
        const q2 = neighbor.dest;
        if (
          q2.reach !== undefined &&
          q.reach !== undefined &&
          q2.reach > q.reach
        ) {
          q.reach = q2.reach;
          changed = true;
        }
      }
      if (qi < quanta.length - 1) {
        const q2 = quanta[qi + 1];
        if (
          q2.reach !== undefined &&
          q.reach !== undefined &&
          q2.reach > q.reach
        ) {
          q.reach = q2.reach;
          changed = true;
        }
      }
      if (changed) {
        changeCount += 1;
        for (let j = 0; j < q.which; j += 1) {
          const q2 = quanta[j];
          if (
            q2.reach !== undefined &&
            q.reach !== undefined &&
            q2.reach < q.reach
          ) {
            q2.reach = q.reach;
          }
        }
      }
    }
    if (changeCount === 0) {
      break;
    }
  }
}

function maxBackwardEdge(q: QuantumBase): number {
  let maxBackward = 0;
  for (const neighbor of q.neighbors) {
    const delta = q.which - neighbor.dest.which;
    if (delta > maxBackward) {
      maxBackward = delta;
    }
  }
  return maxBackward;
}

function findBestLastBeat(
  quanta: QuantumBase[],
  config: JukeboxConfig,
): { index: number; longestReach: number } {
  let longest = 0;
  let longestReach = 0;
  let bestLongIndex = -1;
  let bestLongBack = 0;
  let bestLongReach = 0;
  // Prefer a late beat with strong reachability to avoid early dead-ends.
  for (let i = quanta.length - 1; i >= 0; i -= 1) {
    const q = quanta[i];
    const distanceToEnd = quanta.length - i;
    const reach =
      q.reach !== undefined
        ? ((q.reach - distanceToEnd) * 100) / quanta.length
        : 0;
    if (reach > longestReach && q.neighbors.length > 0) {
      longestReach = reach;
      longest = i;
    }
    const maxBackward = maxBackwardEdge(q);
    if (q.neighbors.length > 0 && maxBackward >= config.minLongBranch) {
      if (i > bestLongIndex) {
        bestLongIndex = i;
        bestLongBack = maxBackward;
        bestLongReach = reach;
      } else if (i === bestLongIndex) {
        if (
          maxBackward > bestLongBack ||
          (maxBackward === bestLongBack && reach > bestLongReach)
        ) {
          bestLongBack = maxBackward;
          bestLongReach = reach;
        }
      }
    }
  }
  if (bestLongIndex >= 0) {
    return { index: bestLongIndex, longestReach: bestLongReach };
  }
  return { index: longest, longestReach };
}

function filterOutBadBranches(quanta: QuantumBase[], lastIndex: number) {
  for (let i = 0; i < lastIndex; i += 1) {
    const q = quanta[i];
    q.neighbors = q.neighbors.filter(
      (neighbor) => neighbor.dest.which < lastIndex,
    );
  }
}

function hasSequentialBranch(
  q: QuantumBase,
  neighbor: Edge,
  lastBranchPoint: number,
) {
  if (q.which === lastBranchPoint) {
    return false;
  }
  const qp = q.prev;
  if (!qp) {
    return false;
  }
  const distance = q.which - neighbor.dest.which;
  for (const prevNeighbor of qp.neighbors) {
    const odistance = qp.which - prevNeighbor.dest.which;
    if (distance === odistance) {
      return true;
    }
  }
  return false;
}

function filterOutSequentialBranches(
  quanta: QuantumBase[],
  lastBranchPoint: number,
) {
  for (let i = quanta.length - 1; i >= 1; i -= 1) {
    const q = quanta[i];
    q.neighbors = q.neighbors.filter(
      (neighbor) => !hasSequentialBranch(q, neighbor, lastBranchPoint),
    );
  }
}

function resolveThreshold(
  quanta: QuantumBase[],
  config: JukeboxConfig,
): number {
  if (config.currentThreshold !== 0) {
    return config.currentThreshold;
  }
  const targetBranchCount = quanta.length / 6;
  for (let t = 10; t < config.maxBranchThreshold; t += 5) {
    const count = collectNearestNeighbors(quanta, t, config);
    if (count >= targetBranchCount) {
      return t;
    }
  }
  return config.maxBranchThreshold;
}

function addAnchorBranch(
  quanta: QuantumBase[],
  threshold: number,
  config: JukeboxConfig,
) {
  if (!config.addLastEdge) {
    return;
  }
  if (longestBackwardBranch(quanta) < 50) {
    insertBestBackwardBranch(quanta, threshold, 65);
  } else {
    insertBestBackwardBranch(quanta, threshold, 55);
  }
}

function applyBranchFilters(
  quanta: QuantumBase[],
  config: JukeboxConfig,
): { lastBranchPoint: number; longestReach: number } {
  calculateReachability(quanta);
  const { index: lastBranchPoint, longestReach } = findBestLastBeat(
    quanta,
    config,
  );
  filterOutBadBranches(quanta, lastBranchPoint);
  if (config.removeSequentialBranches) {
    filterOutSequentialBranches(quanta, lastBranchPoint);
  }
  return { lastBranchPoint, longestReach };
}

export function buildJumpGraph(
  analysis: TrackAnalysis,
  config: JukeboxConfig,
): JukeboxGraphState {
  const quanta = analysis.beats;
  const allEdges: Edge[] = [];
  precalculateNearestNeighbors(
    quanta,
    config.maxBranches,
    config.maxBranchThreshold,
    allEdges,
  );

  const threshold = resolveThreshold(quanta, config);
  collectNearestNeighbors(quanta, threshold, config);
  addAnchorBranch(quanta, threshold, config);
  const { lastBranchPoint, longestReach } = applyBranchFilters(quanta, config);

  return {
    computedThreshold: threshold,
    currentThreshold: threshold,
    lastBranchPoint,
    totalBeats: quanta.length,
    longestReach,
    allEdges,
  };
}
